File: solutions/python/0086-partition-list.py
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        before, after = ListNode(0), ListNode(0) 
        before_curr, after_curr = before, after 

        while head: 
            if head.val < x: 
                before_curr.next, before_curr = head, head 
            else: 
                after_curr.next, after_curr = head, head 
            head = head.next 

        after_curr.next = None 
        before_curr.next = after.next 

        return before.next 

File: solutions/rust/0086-partition-list.rs
impl Solution {
    pub fn partition(mut head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut before = ListNode::new(0); 
        let mut after = ListNode::new(0); 
        let mut before_tail = &mut before; 
        let mut after_tail = &mut after; 

        while let Some(mut node) = head { 
            head = node.next.take(); 
            if node.val < x { 
                before_tail.next = Some(node); 
                before_tail = before_tail.next.as_mut().unwrap(); 
            } else { 
                after_tail.next = Some(node); 
                after_tail = after_tail.next.as_mut().unwrap(); 
            } 
        } 

        before_tail.next = after.next.take(); 

        before.next 
                                                                                                                                                              
    }
}
